File: 6_tree/二分搜索树/285. 二叉搜索树中的中序后继.ts
import { BinaryTree } from '../力扣加加/Tree'
import { deserializeNode } from '../力扣加加/构建类/297二叉树的序列化与反序列化'

/**
 *
 * @param root
 * @param p
 * 给定一棵二叉搜索树root和其中的一个节点 p ，找到该节点在树中的中序后继
 * 如果节点没有中序后继，请返回 null
 */
function inorderSuccessor(root: BinaryTree | null, p: BinaryTree | null): BinaryTree | null {
  if (!root || !p) return null

  let pre: BinaryTree | null = null

  function* inorder(root: BinaryTree | null): Generator<BinaryTree | null> {
    if (!root) return
    yield* inorder(root.left)
    if (pre?.val === p?.val) yield root
    pre = root
    yield* inorder(root.right)
  }

  return inorder(root).next().value || null
}

console.log(inorderSuccessor(deserializeNode([2, 1, 3]), deserializeNode([1])))
console.log(inorderSuccessor(deserializeNode([5, 3, 6, 2, 4, null, null, 1]), deserializeNode([1])))

File: 6_tree/二分搜索树BST/285. 二叉搜索树中的中序后继.py
# Definition for a binary tree node.
from itertools import chain, dropwhile, takewhile
from typing import Generator, Optional


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> Optional['TreeNode']:
        def inorder(node: Optional['TreeNode']) -> Generator['TreeNode', None, None]:
            if not node:
                return
            yield from inorder(node.left)
            yield node
            yield from inorder(node.right)

        gen = dropwhile(lambda node: node.val != p.val, inorder(root))
        next(gen)
        return next(gen, None)

    def inorderSuccessor2(self, root: 'TreeNode', p: 'TreeNode') -> Optional['TreeNode']:
        dfs = lambda node: chain(dfs(node.left), [node], dfs(node.right)) if node else []
        gen = dropwhile(lambda node: node.val != p.val, dfs(root))
        return next(gen, None) if next(gen, None) else None  # 解决了p不在树中的情况
